pkg/xds/generator/proxy_template.go

```diff
@@ -151,30 +153,54 @@ func (_ OutboundProxyGenerator) Generate(ctx xds_context.Context, proxy *model.P
 	resources := make([]*Resource, 0, len(ofaces))
 	names := make(map[string]bool)
 	sourceService := proxy.Dataplane.Spec.GetIdentifyingService()
-	for _, oface := range ofaces {
+	for i, oface := range ofaces {
 		endpoint, err := kuma_mesh.ParseOutboundInterface(oface.Interface)
 		if err != nil {
 			return nil, err
 		}
 
-		serviceTag := kuma_mesh.ServiceTagValue(oface.Service)
-		edsClusterName := outboundClusterName(serviceTag, oface.ServicePort)
-		if used := names[edsClusterName]; !used {
-			resources = append(resources, &Resource{
-				Name:     edsClusterName,
-				Resource: envoy.CreateEdsCluster(ctx, edsClusterName, proxy.Metadata),
-			})
-			resources = append(resources, &Resource{
-				Name:     edsClusterName,
-				Resource: envoy.CreateClusterLoadAssignment(edsClusterName, proxy.OutboundTargets[oface.Service]),
+		route := proxy.TrafficRoutes[oface.Service]
+		if route == nil {
+			return nil, errors.Errorf("outbound interface [%d]{service=%q} has no TrafficRoute", i, oface.Service)
+		}
+
+		var clusters []envoy.ClusterInfo
+		for _, destination := range route.Spec.Conf {
+			service, ok := destination.Destination[kuma_mesh.ServiceTag]
+			if !ok {
+				// all destinations must have `service` tag
+				continue
+			}
+			if destination.Weight == 0 {
+				// Envoy doesn't support 0 weight
+				continue
+			}
+			clusters = append(clusters, envoy.ClusterInfo{
+				Name:   destinationClusterName(service, destination.Destination),
+				Weight: destination.Weight,
+				Tags:   destination.Destination,
 			})
-			names[edsClusterName] = true
+		}
+
+		for _, cluster := range clusters {
+			if used := names[cluster.Name]; !used {
+				resources = append(resources, &Resource{
+					Name:     cluster.Name,
+					Resource: envoy.CreateEdsCluster(ctx, cluster.Name, proxy.Metadata),
+				})
+				endpoints := model.EndpointList(proxy.OutboundTargets[cluster.Tags[kuma_mesh.ServiceTag]]).Filter(kuma_mesh.MatchTags(cluster.Tags))
+				resources = append(resources, &Resource{
+					Name:     cluster.Name,
+					Resource: envoy.CreateClusterLoadAssignment(cluster.Name, endpoints),
+				})
+				names[cluster.Name] = true
+			}
 		}
 
 		outboundListenerName := fmt.Sprintf("outbound:%s:%d", endpoint.DataplaneIP, endpoint.DataplanePort)
 		if used := names[outboundListenerName]; !used {
```

Comment:
btw. why do we need this? Is there a case when listener names are the same?

Reply:
yes. Since previously we didn't have resource validation, `Dataplane` could be extremely weird

Reply:
I've refactored it in a nicer `ResourceSet` abstraction